File: backend/src/models/customer.model.js
const { pool } = require('../config/database');

class CustomerModel {
  static async create(data, createdBy = null) {
    const result = await pool.query(`
      INSERT INTO customers (
        name, phone_number, address, whatsapp_number, location_link,
        latitude, longitude, permanent_quantity, sub_area_id,
        sort_number, is_approved, pending_approval, is_active, shift
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
      RETURNING *
    `, [
      data.name,
      data.phone_number,
      data.address,
      data.whatsapp_number || null,
      data.location_link || null,
      data.latitude || null,
      data.longitude || null,
      data.permanent_quantity || 0,
      data.sub_area_id,
      data.sort_number ? parseFloat(data.sort_number) : null,
      false,
      true,
      data.is_active !== undefined ? data.is_active : true,
      data.shift || null
    ]);

    return result.rows[0];
  }

  static async findById(id) {
    const result = await pool.query(`
      SELECT 
        c.*,
        sa.name as sub_area_name,
        a.name as area_name,
        COALESCE(
          (SELECT SUM(e.milk_quantity * e.rate - e.collected_money) 
           FROM entries e 
           WHERE e.customer_id = c.id),
          0
        ) as total_pending_money,
        COALESCE(
          (SELECT e.pending_bottles 
           FROM entries e 
           WHERE e.customer_id = c.id 
           ORDER BY e.entry_date DESC 
           LIMIT 1),
          0
        ) as last_time_pending_bottles
      FROM customers c
      LEFT JOIN sub_areas sa ON c.sub_area_id = sa.id
      LEFT JOIN areas a ON sa.area_id = a.id
      WHERE c.id = $1
    `, [id]);
    return result.rows[0];
  }

  static async getByDeliveryBoy(deliveryBoyId, filters = {}) {
    let query = `
      SELECT 
        c.*,
        sa.name as sub_area_name,
        a.name as area_name,
        COALESCE(
          (SELECT SUM(e.milk_quantity * e.rate - e.collected_money) 
           FROM entries e 
           WHERE e.customer_id = c.id),
          0
        ) as total_pending_money,
        COALESCE(
          (SELECT e.pending_bottles 
           FROM entries e 
           WHERE e.customer_id = c.id 
           ORDER BY e.entry_date DESC 
           LIMIT 1),
          0
        ) as last_time_pending_bottles,
        COALESCE(
          (SELECT e.is_delivered 
           FROM entries e 
           WHERE e.customer_id = c.id 
           AND e.entry_date = CURRENT_DATE
           ORDER BY e.entry_date DESC 
           LIMIT 1),
          null
        ) as today_delivery_status
      FROM customers c
      LEFT JOIN sub_areas sa ON c.sub_area_id = sa.id
      LEFT JOIN areas a ON sa.area_id = a.id
      WHERE c.sub_area_id IN (
        SELECT sub_area_id FROM delivery_boy_subareas WHERE delivery_boy_id = $1
      ) AND c.is_approved = true AND c.is_active = true
    `;

    const params = [deliveryBoyId];
    let paramCount = 2;

    if (filters.area_id) {
      query += ` AND a.id = $${paramCount}`;
      params.push(filters.area_id);
      paramCount++;
    }

    if (filters.sub_area_id) {
      query += ` AND c.sub_area_id = $${paramCount}`;
      params.push(filters.sub_area_id);
      paramCount++;
    }

    if (filters.delivery_status) {
      if (filters.delivery_status === 'delivered') {
        query += ` AND EXISTS (
          SELECT 1 FROM entries e 
          WHERE e.customer_id = c.id 
          AND e.entry_date = CURRENT_DATE 
          AND e.is_delivered = true
        )`;
      } else if (filters.delivery_status === 'pending') {
        query += ` AND NOT EXISTS (
          SELECT 1 FROM entries e 
          WHERE e.customer_id = c.id 
          AND e.entry_date = CURRENT_DATE
        )`;
      } else if (filters.delivery_status === 'notDelivered') {
        query += ` AND EXISTS (
          SELECT 1 FROM entries e 
          WHERE e.customer_id = c.id 
          AND e.entry_date = CURRENT_DATE 
          AND e.is_delivered = false
        )`;
      }
    }

    if (filters.search) {
      query += ` AND (c.name ILIKE $${paramCount} OR c.phone_number ILIKE $${paramCount})`;
      params.push(`%${filters.search}%`);
      paramCount++;
    }

    query += ' ORDER BY c.sort_number ASC, c.created_at DESC';

    const result = await pool.query(query, params);
    return result.rows;
  }

  static async update(id, data) {
    const fields = [];
    const values = [];
    let paramCount = 1;

    Object.keys(data).forEach(key => {
      if (data[key] !== undefined && key !== 'id') {
        fields.push(`${key} = $${paramCount}`);
        values.push(data[key]);
        paramCount++;
      }
    });

    if (fields.length === 0) return null;

    values.push(id);
    const query = `
      UPDATE customers 
      SET ${fields.join(', ')}
      WHERE id = $${paramCount}
      RETURNING *
    `;

    const result = await pool.query(query, values);
    return result.rows[0];
  }

  static async delete(id) {
    const result = await pool.query(
      'DELETE FROM customers WHERE id = $1 RETURNING *',
      [id]
    );
    return result.rows[0];
  }

  static async getCustomerEntries(customerId, startDate = null, endDate = null) {
    let query = `
      SELECT * FROM entries 
      WHERE customer_id = $1
    `;

    const params = [customerId];
    let paramCount = 2;

    if (startDate) {
      query += ` AND entry_date >= $${paramCount}`;
      params.push(startDate);
      paramCount++;
    }

    if (endDate) {
      query += ` AND entry_date <= $${paramCount}`;
      params.push(endDate);
      paramCount++;
    }

    query += ' ORDER BY entry_date DESC';

    const result = await pool.query(query, params);
    return result.rows;
  }
}

module.exports = CustomerModel;

File: delivery_app/lib/data/models/customer_model.dart
import 'package:equatable/equatable.dart';

class CustomerModel extends Equatable {
  final int id;
  final String name;
  final String phoneNumber;
  final String? address;
  final String? whatsappNumber;
  final String? locationLink;
  final double? latitude;
  final double? longitude;
  final double permanentQuantity;
  final int? subAreaId;
  final String? subAreaName;
  final String? areaName;
  final int? deliveryBoyId;
  final String? deliveryBoyName;
  final double? sortNumber;
  final bool isApproved;
  final bool pendingApproval;
  final bool? isActive;
  final String? shift;
  final double? totalPendingMoney;
  final int? lastTimePendingBottles;
  final bool? todayDeliveryStatus;

  const CustomerModel({
    required this.id,
    required this.name,
    required this.phoneNumber,
    this.address,
    this.whatsappNumber,
    this.locationLink,
    this.latitude,
    this.longitude,
    required this.permanentQuantity,
    this.subAreaId,
    this.subAreaName,
    this.areaName,
    this.deliveryBoyId,
    this.deliveryBoyName,
    this.sortNumber,
    required this.isApproved,
    required this.pendingApproval,
    this.isActive,
    this.shift,
    this.totalPendingMoney,
    this.lastTimePendingBottles,
    this.todayDeliveryStatus,
  });

  factory CustomerModel.fromJson(Map<String, dynamic> json) {
    return CustomerModel(
      id: json['id'],
      name: json['name'],
      phoneNumber: json['phone_number'],
      address: json['address'],
      whatsappNumber: json['whatsapp_number'],
      locationLink: json['location_link'],
      latitude: json['latitude'] != null
          ? double.tryParse(json['latitude'].toString())
          : null,
      longitude: json['longitude'] != null
          ? double.tryParse(json['longitude'].toString())
          : null,
      permanentQuantity:
          double.tryParse(json['permanent_quantity'].toString()) ?? 0,
      subAreaId: json['sub_area_id'],
      subAreaName: json['sub_area_name'],
      areaName: json['area_name'],
      deliveryBoyId: json['delivery_boy_id'],
      deliveryBoyName: json['delivery_boy_name'],
      sortNumber: json['sort_number'] != null
          ? (json['sort_number'] is num
                ? (json['sort_number'] as num).toDouble()
                : double.tryParse(json['sort_number'].toString()))
          : null,
      isApproved: json['is_approved'] ?? false,
      pendingApproval: json['pending_approval'] ?? false,
      isActive: json['is_active'] != null
          ? (json['is_active'] is bool
                ? json['is_active']
                : (json['is_active'].toString() == '1' ||
                      json['is_active'].toString().toLowerCase() == 'true'))
          : true,
      shift: json['shift'] ?? json['shift_time'] ?? null,
      totalPendingMoney: json['total_pending_money'] != null
          ? double.tryParse(json['total_pending_money'].toString())
          : null,
      lastTimePendingBottles: json['last_time_pending_bottles'],
      todayDeliveryStatus: json['today_delivery_status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'phone_number': phoneNumber,
      'address': address,
      'whatsapp_number': whatsappNumber,
      'location_link': locationLink,
      'latitude': latitude,
      'longitude': longitude,
      'permanent_quantity': permanentQuantity,
      'sub_area_id': subAreaId,
      'delivery_boy_id': deliveryBoyId,
      'sort_number': sortNumber,
      'is_approved': isApproved,
      'pending_approval': pendingApproval,
      'is_active': isActive,
      'shift': shift,
    };
  }

  @override
  List<Object?> get props => [
    id,
    name,
    phoneNumber,
    address,
    whatsappNumber,
    locationLink,
    latitude,
    longitude,
    permanentQuantity,
    subAreaId,
    subAreaName,
    areaName,
    deliveryBoyId,
    deliveryBoyName,
    sortNumber,
    isApproved,
    pendingApproval,
    totalPendingMoney,
    lastTimePendingBottles,
    todayDeliveryStatus,
  ];
}
